src/utils/encoding/cid_de_cbor.rs

```diff
@@ -0,0 +1,179 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use cid::serde::BytesToCidVisitor;
+use cid::Cid;
+use core::fmt;
+use serde::de::{self, DeserializeSeed, SeqAccess, Visitor};
+use serde::Deserializer;
+use std::ops::{Deref, DerefMut};
+
+/// [`CidVec`] allows for efficient zero-copy de-serialization of `DAG_CBOR`-encoded nodes into a
+/// vector of [`Cid`].
+#[derive(Default)]
+pub struct CidVec(Vec<Cid>);
+
+impl Deref for CidVec {
+    type Target = Vec<Cid>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl DerefMut for CidVec {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+// [`CollectCid`] struct allows for recursive traversal of [`Ipld`] in order to retrieve all the
+// CIDs it encounters, using a preallocated vector.
+struct CollectCid<'a>(&'a mut Vec<Cid>);
+
+impl<'de, 'a> DeserializeSeed<'de> for CollectCid<'a> {
+    type Value = ();
+
+    fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        struct CollectCidVisitor<'a>(&'a mut Vec<Cid>);
+
+        impl<'de, 'a> Visitor<'de> for CollectCidVisitor<'a> {
+            type Value = ();
+
+            fn expecting(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+                fmt.write_str("any valid IPLD kind")
+            }
+
+            #[inline]
+            fn visit_str<E>(self, _value: &str) -> Result<Self::Value, E>
```

Comment:
nonblocking: can you macro these away with a comment? Or at least lift the most relevant fns to the top?
"the default visitor errors when visiting these, but we want to skip over them, so return `Ok(())`"
Out of curiosity why have you only done that for _some_ of the visited types?

Reply:
We only care about list, map and newtype_struct, because a list and a map could contain more of those and the only type that converts to a CID is the newtype_struct.

Reply:
I don't think that introducing a macro here is going to help anything, it's just going to obfuscate the obvious. I'll re-arrange the order, however.